data_model/src/role.rs

```diff
@@ -7,67 +7,38 @@ use std::collections::btree_set;
 
 use derive_more::{Constructor, Display, FromStr};
 use getset::Getters;
-use iroha_data_model_derive::IdOrdEqHash;
-use iroha_ffi::FfiType;
+use iroha_data_model_derive::IdEqOrdHash;
 use iroha_schema::IntoSchema;
 use parity_scale_codec::{Decode, Encode};
 use serde::{Deserialize, Serialize};
 
 use crate::{
-    ffi::declare_item,
+    ffi,
     permission::{Permissions, Token as PermissionToken},
     Identifiable, Name, Registered,
 };
 
 /// Collection of [`RoleId`](Id)s
 pub type RoleIds = btree_set::BTreeSet<<Role as Identifiable>::Id>;
 
-/// Identification of a role.
-#[derive(
-    Debug,
-    Display,
-    Constructor,
-    FromStr,
-    Clone,
-    PartialEq,
-    Eq,
-    PartialOrd,
-    Ord,
-    Hash,
-    Decode,
-    Encode,
-    Deserialize,
-    Serialize,
-    FfiType,
-    IntoSchema,
-)]
-#[repr(transparent)]
-#[serde(transparent)]
-// SAFETY: RoleId has no trap representations in Name
-#[ffi_type(unsafe {robust})]
-pub struct Id {
-    /// Role name, should be unique .
-    pub name: Name,
+ffi::ffi_item! {
+    /// Identification of a role.
+    #[derive(Debug, Display, Clone, Constructor, FromStr, PartialEq, Eq, PartialOrd, Ord, Hash, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+    #[serde(transparent)]
+    #[repr(transparent)]
+    pub struct Id {
+        /// Role name, should be unique .
+        pub name: Name,
+    }
+
+    ffi_type(opaque)
 }
 
-declare_item! {
+ffi::ffi_item! {
     /// Role is a tag for a set of permission tokens.
-    #[derive(
-        Debug,
-        Display,
-        Clone,
-        IdOrdEqHash,
-        Getters,
-        Decode,
-        Encode,
-        Deserialize,
-        Serialize,
-        FfiType,
-        IntoSchema,
-    )]
-    #[cfg_attr(all(feature = "ffi_export", not(feature = "ffi_import")), iroha_ffi::ffi_export)]
-    #[cfg_attr(feature = "ffi_import", iroha_ffi::ffi_import)]
```

Comment:
this is now automatically derived either with `FfiType` (added on `feature = "ffi_export`) or `iroha_ffi::ffi!` (added on `feature = "ffi_import"`)